source/common/ssl/context_impl.cc

```diff
@@ -131,6 +131,15 @@ ContextImpl::ContextImpl(ContextManagerImpl& parent, Stats::Scope& scope,
     verify_mode = SSL_VERIFY_PEER | SSL_VERIFY_FAIL_IF_NO_PEER_CERT;
   }
 
+  // NOTE: We're using SSL_CTX_set_cert_verify_callback() instead of X509_verify_cert()
+  // directly. However, our new callback is still calling X509_verify_cert() under
+  // the hood. Therefore, to ignore cert expiration, we need to set the callback
+  // for X509_verify_cert to ignore that error.
+  if (config.allowExpiredCertificate()) {
```

Comment:
Actually, this should be inside the `if (!config.caCert().empty())` block, since this makes sense only when `trusted_ca` is configured (see: https://github.com/envoyproxy/envoy/pull/3549, https://github.com/envoyproxy/envoy/pull/3550).

Reply:
Doesn't it also make sense in the `verify_certificate_hash` case without a CA per #3549 ?

How about if I move it into the`if (verify_mode != SSL_VERIFY_NONE)` block where it sets `verifyCallback`?

Reply:
No, it doesn't, since validity period is a property of CA's signature.

Both: `verify_certificate_spki` and `verify_certificate_hash` verify only hash of SPKI / certificate, when used without `trusted_ca`.